File: lintcode/median-of-two-sorted-arrays.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/median-of-two-sorted-arrays/

class Solution:
    def find_kth(self, A, B, k):
        if len(A) > len(B):
            return self.find_kth(B, A, k)
        if len(A) == 0:
            return B[k-1]
        if k == 1:
            return min(A[0], B[0])

        pa = min(k/2, len(A))
        pb = k-pa
        if A[pa-1] < B[pb-1]:
            return self.find_kth(A[pa:], B, k-pa)
        elif A[pa-1] > B[pb-1]:
            return self.find_kth(A, B[pb:], k-pb)
        else:
            return A[pa-1]

    """
    @param A: An integer array.
    @param B: An integer array.
    @return: a double whose format is *.5 or *.0
    """
    def findMedianSortedArrays(self, A, B):
        # write your code here
        total = len(A) + len(B)
        if total % 2 == 1:
            return self.find_kth(A, B, total/2+1)
        else:
            return (self.find_kth(A, B, total/2) + self.find_kth(A, B, total/2+1)) * 1.0 / 2

if __name__ == '__main__':
    s = Solution()
    print s.findMedianSortedArrays([1,2,3,4,5,6], [2,3,4,5])

File: leetcode/median-of-two-sorted-arrays.cc
// Question: https://leetcode.com/problems/median-of-two-sorted-arrays/

#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int sum = nums1.size() + nums2.size();
        if (sum % 2 == 0) {
            return (find_kth(nums1, nums2, sum/2) + find_kth(nums1, nums2, sum/2+1)) * 1.0 / 2;
        } else {
            return find_kth(nums1, nums2, sum/2+1);
        }
    }

private:
    int find_kth(vector<int> nums1, vector<int> nums2, int k) {
        if (nums1.size() > nums2.size()) {
            return find_kth(nums2, nums1, k);
        }
        if (nums1.size() == 0) {
            return nums2.at(k-1);
        }
        if (k == 1) {
            return nums1.at(0) < nums2.at(0) ? nums1.at(0) : nums2.at(0);
        }

        int p1 = k/2 < nums1.size() ? k/2 : nums1.size();
        int p2 = k-p1;
        if (nums1.at(p1-1) > nums2.at(p2-1)) {
            return find_kth(nums1, vector<int>(nums2.begin()+p2, nums2.end()), k-p2);
        } else if (nums1.at(p1-1) < nums2.at(p2-1)) {
            return find_kth(vector<int>(nums1.begin()+p1, nums1.end()), nums2, k-p1);
        } else {
            return nums1.at(p1-1);
        }
    }
};

int main(int argc, char** argv) {
    int n1[] = {3,5,7};
    int n2[] = {2};
    vector<int> nums1(n1, n1 + sizeof(n1) / sizeof(int));
    vector<int> nums2(n2, n2 + sizeof(n2) / sizeof(int));
    Solution s;
    cout << s.findMedianSortedArrays(nums1, nums2) << endl;
}
